File: Trie/208_implement_trie/208.rs
use std::collections::HashMap;


#[derive(Debug)]
struct TrieNode {
    next_nodes: HashMap<char, TrieNode >,
    end: bool,
}

impl TrieNode {
    fn new() -> Self {
        return TrieNode { next_nodes: HashMap::new(), end: false };
    }
}

struct Trie {
    root: TrieNode,
}

pub impl Trie {
    
    fn new() -> Self {
        return Trie {root: TrieNode::new()};
    }

    fn insert(&mut self, word: String) {
        let mut current_node = &mut self.root;

        for c in word.chars() {
            let next_node = current_node.next_nodes.entry(c).or_insert(TrieNode::new());
            current_node = next_node;
        }
        current_node.end = true;
    }

    fn search(&self, word: String) -> bool {
        let mut current_node = &self.root;
        
        for c in word.chars() {
            match current_node.next_nodes.get(&c) {
                Some(next_node) => current_node = next_node,
                None => return false,
            }
        }
        return current_node.end;
    }

    fn starts_with(&self, prefix: String) -> bool {
        let mut current_node = &self.root;

        for c in prefix.chars() {
            match current_node.next_nodes.get(&c) {
                Some(next_node) => current_node = next_node,
                None => return false,
            }
        }
        return true;
    }
}

File: Trie/208_implement_trie/208.java
class TrieNode {

    TrieNode[] root;
    private static final int letters = 26;
    boolean isWord;

    public TrieNode() {
        root = new TrieNode[letters];
        isWord = false;
    }

    public boolean containsKey(char ch) {
        return root[ch - 'a'] != null;
    }

    public TrieNode insertKey(char ch) {
        TrieNode newNode = new TrieNode();
        root[ch - 'a'] = newNode;
        return newNode;
    }

    public TrieNode getKey(char ch) {
        return root[ch - 'a'];
    }

    public boolean isWord() {
        return isWord;
    }

    public void setWord() {
        isWord = true;
    }
}

class Trie {

    TrieNode root;

    public Trie() {
        root = new TrieNode();
    }

    public void insert(String word) {
        TrieNode curNode = root;

        for (int i = 0; i < word.length(); i++) {
            char curr = word.charAt(i);

            if (!curNode.containsKey(curr)) {
                curNode = curNode.insertKey(curr);
            } else {
                curNode = curNode.getKey(curr);
            }
        }

        curNode.setWord();
    }

    public boolean search(String word) {
        TrieNode search = searchWord(word);
        return (search != null && search.isWord());
    }

    public boolean startsWith(String prefix) {
        TrieNode search = searchWord(prefix);
        return search != null;
    }

    private TrieNode searchWord(String word) {

        TrieNode curNode = root;

        for (int i = 0; i < word.length(); i++) {
            char curr = word.charAt(i);

            if (curNode.containsKey(curr)) {
                curNode = curNode.getKey(curr);
            } else {
                return null;
            }
        }

        return curNode;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */